File: backend/ai/scheduler.py
"""24-hour analysis trigger logic.

While the app is running, periodically checks if it's time
to run the daily AI analysis pipeline.
"""

import asyncio
import logging
from datetime import datetime, timedelta

from sqlalchemy import select

from backend.database.connection import AsyncSessionLocal
from backend.database.models import SystemState

logger = logging.getLogger(__name__)


class AnalysisScheduler:
    """Background scheduler that triggers AI analysis every 24 hours.

    Checks hourly whether 24h has passed since the last analysis.
    This handles the case where the app is left open for days.
    """

    def __init__(self):
        self._task: asyncio.Task | None = None
        self._running = False

    async def start(self):
        """Start the background scheduler loop."""
        if self._running:
            return
        self._running = True
        self._task = asyncio.create_task(self._loop())
        logger.info("Analysis scheduler started")

    async def _loop(self):
        """Check every hour if analysis is needed."""
        while self._running:
            await asyncio.sleep(3600)  # Check every hour
            try:
                await self._check_and_trigger()
            except Exception as e:
                logger.error("Scheduler check failed: %s", e)

    async def _check_and_trigger(self):
        """Check if 24h has passed since last analysis and trigger if needed."""
        async with AsyncSessionLocal() as session:
            result = await session.execute(
                select(SystemState).where(SystemState.key == "last_ai_analysis")
            )
            state = result.scalar_one_or_none()

            now = datetime.utcnow()
            should_run = True

            if state and state.value:
                try:
                    last_run = datetime.fromisoformat(state.value)
                    should_run = (now - last_run) > timedelta(hours=24)
                except (ValueError, TypeError):
                    pass

            if should_run:
                logger.info("24h threshold reached, triggering daily pipeline")
                try:
                    from backend.services.daily_pipeline import DailyPipeline

                    pipeline = DailyPipeline()
                    report = await pipeline.run_full(session)

                    if report.success:
                        logger.info(
                            "Scheduled pipeline completed successfully: %d steps",
                            len(report.steps),
                        )
                    else:
                        logger.error(
                            "Scheduled pipeline failed: %s", report.error
                        )
                except Exception as e:
                    logger.error("Failed to run daily pipeline: %s", e)

                    # Even on failure, update the timestamp to avoid rapid retries
                    if state:
                        state.value = now.isoformat()
                        state.updated_at = now
                    else:
                        session.add(
                            SystemState(
                                key="last_ai_analysis",
                                value=now.isoformat(),
                                updated_at=now,
                            )
                        )
                    await session.commit()

    async def stop(self):
        """Stop the background scheduler."""
        self._running = False
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None
        logger.info("Analysis scheduler stopped")

File: src-tauri/src/server/ai/scheduler.rs
use std::sync::Arc;
use std::sync::atomic::Ordering;
use tokio::time::{Duration, interval};
use crate::server::AppState;
use crate::server::services::daily_pipeline;

pub fn start(state: Arc<AppState>) {
    tokio::spawn(async move {
        // Check every hour
        let mut tick = interval(Duration::from_secs(3600));
        tick.tick().await; // First tick is immediate, skip it
        loop {
            tick.tick().await;
            if let Err(e) = check_and_trigger(&state).await {
                tracing::error!("Scheduler check failed: {}", e);
            }
        }
    });
    tracing::info!("Background analysis scheduler started (24h cycle, hourly checks)");
}

async fn check_and_trigger(state: &Arc<AppState>) -> anyhow::Result<()> {
    if state.analysis_running.load(Ordering::Relaxed) {
        tracing::debug!("Analysis already running, skipping scheduler check");
        return Ok(());
    }

    if state.settings.anthropic_api_key.is_empty() {
        return Ok(());
    }

    let last_analysis: Option<(Option<String>,)> = sqlx::query_as(
        "SELECT value FROM system_state WHERE key = 'last_ai_analysis'"
    ).fetch_optional(&state.db).await?;

    let should_run = match last_analysis {
        None => true,
        Some((None,)) => true,
        Some((Some(ts),)) => {
            if let Ok(last) = chrono::NaiveDateTime::parse_from_str(&ts, "%Y-%m-%d %H:%M:%S") {
                let elapsed = chrono::Utc::now().naive_utc() - last;
                elapsed.num_hours() >= 24
            } else {
                true
            }
        }
    };

    if should_run {
        tracing::info!("Scheduler triggering daily analysis pipeline");
        let state = state.clone();
        tokio::spawn(async move {
            if let Err(e) = daily_pipeline::run_full(&state).await {
                tracing::error!("Daily pipeline failed: {}", e);
            }
        });
    }

    Ok(())
}
